File: backend/app/schemas/viewing_request.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class ViewingRequestCreate(BaseModel):
    message: str | None = Field(default=None, max_length=500)


class ViewingRequestPropertySummary(BaseModel):
    id: int
    title: str
    city: str | None
    image_url: str | None
    status: str

    model_config = ConfigDict(from_attributes=True)


class ViewingRequestRenterResponse(BaseModel):
    id: int
    property_id: int
    property: ViewingRequestPropertySummary
    status: str
    message: str | None
    created_at: datetime
    updated_at: datetime
    responded_at: datetime | None


class ViewingRequestRealtorResponse(BaseModel):
    id: int
    property_id: int
    property: ViewingRequestPropertySummary
    requester_id: int
    requester_email: str
    status: str
    message: str | None
    created_at: datetime
    updated_at: datetime
    responded_at: datetime | None


class ViewingRequestRenterListResponse(BaseModel):
    items: list[ViewingRequestRenterResponse]
    total: int
    limit: int
    offset: int


class ViewingRequestRealtorListResponse(BaseModel):
    items: list[ViewingRequestRealtorResponse]
    total: int
    limit: int
    offset: int

File: frontend/types/viewingRequest.ts
export type ViewingRequestStatus =
  | "pending"
  | "accepted"
  | "declined"
  | "cancelled"

export type ViewingRequestPropertySummary = {
  id: number
  title: string
  city: string | null
  image_url: string | null
  status: string
}

export type ViewingRequest = {
  id: number
  property_id: number
  property: ViewingRequestPropertySummary
  status: ViewingRequestStatus
  message: string | null
  created_at: string
  updated_at: string
  responded_at: string | null
}

export type ViewingRequestRealtor = ViewingRequest & {
  requester_id: number
  requester_email: string
}

export type ViewingRequestListResponse = {
  items: ViewingRequest[]
  total: number
  limit: number
  offset: number
}

export type ViewingRequestRealtorListResponse = {
  items: ViewingRequestRealtor[]
  total: number
  limit: number
  offset: number
}

export function getViewingRequestStatusLabel(status: ViewingRequestStatus): string {
  switch (status) {
    case "pending":
      return "Waiting for realtor response"
    case "accepted":
      return "Accepted"
    case "declined":
      return "Declined"
    case "cancelled":
      return "Cancelled"
    default:
      return status
  }
}

export function isActiveViewingRequest(status: ViewingRequestStatus): boolean {
  return status === "pending" || status === "accepted"
}

export function canCancelViewingRequest(status: ViewingRequestStatus): boolean {
  return isActiveViewingRequest(status)
}

export function isPropertyPubliclyAvailable(
  propertyStatus: string | null | undefined
): boolean {
  return propertyStatus === "available"
}
